modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/database/H2TreeIndex.java

```diff
@@ -528,23 +532,15 @@ private static boolean isExpired(@NotNull H2Row row) {
         return row.expireTime() > 0 && row.expireTime() <= U.currentTimeMillis();
     }
 
-    /** {@inheritDoc} */
-    @Override public void destroy(boolean rmvIdx) {
-        destroy0(rmvIdx, false);
-    }
-
-    /** {@inheritDoc} */
-    @Override public void asyncDestroy(boolean rmvIdx) {
-        destroy0(rmvIdx, true);
-    }
-
     /**
-     * Internal method for destroying index with async option.
+     * Internal method for destroying index. For {@link H2TreeIndex} destroy operation is asynchronous.
      *
      * @param rmvIdx Flag remove.
-     * @param async Destroy asynchronously.
      */
-    private void destroy0(boolean rmvIdx, boolean async) {
+    @Override public void destroy(boolean rmvIdx) {
+        if (!markDestroyed())
+            return;
+
         try {
             if (cctx.affinityNode() && rmvIdx) {
                 assert cctx.shared().database().checkpointLockIsHeldByThread();
```

Comment:
Why we expects checkpoint is held by thread?

Reply:
Read lock moved inside of #destroy method, to delete all index segments from metatree under single read lock.